contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,823 @@
+//! Extension of ERC-1155 that adds tracking of total supply per token id.
+//!
+//! Useful for scenarios where Fungible and Non-fungible tokens have to be
+//! clearly identified. Note: While a `_total_supply` of 1 might mean the
+//! corresponding is an NFT, there are no guarantees that no other tokens
+//! with the same id are not going to be minted.
+//!
+//! NOTE: This contract implies a global limit of 2**256 - 1 to the number
+//! of tokens that can be minted.
+//!
+//! CAUTION: This extension should not be added in an upgrade to an already
+//! deployed contract.
+
+use alloc::{vec, vec::Vec};
+
+use alloy_primitives::{Address, U256};
+use openzeppelin_stylus_proc::interface_id;
+use stylus_sdk::{
+    abi::Bytes,
+    msg,
+    prelude::{public, sol_storage},
+    storage::TopLevelStorage,
+};
+
+use crate::{
+    token::erc1155::{self, Erc1155, IErc1155},
+    utils::math::storage::SubAssignUnchecked,
+};
+
+sol_storage! {
+    /// State of an [`Erc1155Supply`] token.
+    pub struct Erc1155Supply {
+        /// ERC-1155 contract storage.
+        Erc1155 erc1155;
+        /// Mapping from token id to total supply.
+        mapping(uint256 => uint256) _total_supply;
+        /// Total supply of all token ids.
+        uint256 _total_supply_all;
+    }
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc1155Supply {}
+
+/// Required interface of a [`Erc1155Supply`] contract.
+#[interface_id]
+pub trait IErc1155Supply {
```

Comment:
I have a feeling, that we probably don't need a trait for that kind of implementation. Just reexport all functions withing a simple `impl` block. I don't understand use-case of `IErc1155Supply` trait outside of this module

Reply:
This relates to our [discussion in group chat](https://openzeppelin.slack.com/archives/C06R5AGFE9G/p1733325243812519). 

@qalisander what do you say we leave the trait in for now, until we get an agreement on how we want to handle them?  If we decide to remove some traits, we can remove `IErc1155Supply` then.

@bidzyyys pinging you to take not of this trait and Sasha's comment, so that we're on the same page for the discussion.

Reply:
Looks like resolved.